File: mage/src/main/java/mil/nga/giat/mage/login/LoginActivity.java
package mil.nga.giat.mage.login;

import android.app.Activity;
import android.app.Dialog;
import android.content.DialogInterface;
import android.content.DialogInterface.OnClickListener;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Typeface;
import android.net.Uri;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.MotionEvent;
import android.view.View;
import android.view.View.OnTouchListener;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.ViewModelProvider;
import androidx.work.WorkManager;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GooglePlayServicesUtil;

import org.apache.commons.lang3.StringUtils;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Iterator;
import java.util.Map;
import java.util.TreeMap;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;
import mil.nga.giat.mage.LandingActivity;
import mil.nga.giat.mage.MageApplication;
import mil.nga.giat.mage.R;
import mil.nga.giat.mage._server5.login.SignupActivity_server5;
import mil.nga.giat.mage.cache.CacheUtils;
import mil.nga.giat.mage.disclaimer.DisclaimerActivity;
import mil.nga.giat.mage.event.EventsActivity;
import mil.nga.giat.mage.login.LoginViewModel.Authentication;
import mil.nga.giat.mage.login.idp.IdpLoginFragment;
import mil.nga.giat.mage.login.ldap.LdapLoginFragment;
import mil.nga.giat.mage.login.mage.MageLoginFragment;
import mil.nga.giat.mage.sdk.Compatibility;
import mil.nga.giat.mage.sdk.datastore.DaoStore;
import mil.nga.giat.mage.sdk.datastore.user.Event;
import mil.nga.giat.mage.sdk.datastore.user.User;
import mil.nga.giat.mage.sdk.datastore.user.UserHelper;
import mil.nga.giat.mage.sdk.exceptions.UserException;
import mil.nga.giat.mage.sdk.login.AuthenticationStatus;
import mil.nga.giat.mage.sdk.login.AuthorizationStatus;
import mil.nga.giat.mage.sdk.preferences.PreferenceHelper;
import mil.nga.giat.mage.sdk.utils.MediaUtility;
import mil.nga.giat.mage.sdk.utils.UserUtility;

@AndroidEntryPoint
public class LoginActivity extends AppCompatActivity {

   public static final String EXTRA_CONTINUE_SESSION = "CONTINUE_SESSION";
   public static final String EXTRA_CONTINUE_SESSION_WHILE_USING = "CONTINUE_SESSION_WHILE_USING";

   @Inject
   protected MageApplication application;

   @Inject
   protected SharedPreferences preferences;

   private LoginViewModel viewModel;

   private TextView mServerURL;

   private String mOpenFilePath;

   private boolean mContinueSession;

   public final TextView getServerUrlText() {
      return mServerURL;
   }

   @Override
   public void onCreate(Bundle savedInstanceState) {
      super.onCreate(savedInstanceState);

      Intent intent = getIntent();

      mContinueSession = getIntent().getBooleanExtra(EXTRA_CONTINUE_SESSION, false);

      boolean continueSessionWhileUsing = getIntent().getBooleanExtra(EXTRA_CONTINUE_SESSION_WHILE_USING, false);
      intent.removeExtra(EXTRA_CONTINUE_SESSION_WHILE_USING);
      if (continueSessionWhileUsing && savedInstanceState == null) {
         showSessionExpiredDialog();
      }

      if (intent.getBooleanExtra("LOGOUT", false)) {
         application.onLogout(true, null);
      }

      // IMPORTANT: load the configuration from preferences files and server
      PreferenceHelper preferenceHelper = PreferenceHelper.getInstance(getApplicationContext());
      // TODO make sure sdk config is loaded
//      preferenceHelper.initialize(false, mil.nga.giat.mage.sdk.R.xml.class, R.xml.class);
      preferenceHelper.initialize(false, R.xml.class);

      // check if the database needs to be upgraded, and if so log them out
      if (DaoStore.DATABASE_VERSION != preferences.getInt(getResources().getString(R.string.databaseVersionKey), 0)) {
         application.onLogout(true, null);
      }

      preferences.edit().putInt(getString(R.string.databaseVersionKey), DaoStore.DATABASE_VERSION).apply();

      // check google play services version
      int isGooglePlayServicesAvailable = GooglePlayServicesUtil.isGooglePlayServicesAvailable(getApplicationContext());
      if (isGooglePlayServicesAvailable != ConnectionResult.SUCCESS) {
         if (GooglePlayServicesUtil.isUserRecoverableError(isGooglePlayServicesAvailable)) {
            Dialog dialog = GooglePlayServicesUtil.getErrorDialog(isGooglePlayServicesAvailable, this, 1);
            dialog.setOnCancelListener(new DialogInterface.OnCancelListener() {
               @Override
               public void onCancel(DialogInterface dialog) {
                  dialog.dismiss();
                  finish();
               }
            });
            dialog.show();
         } else {
            new AlertDialog.Builder(this).setTitle("Google Play Services").setMessage("Google Play Services is not installed, or needs to be updated.  Please update Google Play Services before continuing.").setPositiveButton(android.R.string.ok, new OnClickListener() {
               @Override
               public void onClick(DialogInterface dialog, int which) {
                  dialog.dismiss();
                  finish();
               }
            }).show();
         }
      }

      // Handle when MAGE was launched with a Uri (such as a local or remote cache file)
      Uri uri = intent.getData();
      if (uri == null) {
         Bundle bundle = intent.getExtras();
         if (bundle != null) {
            Object objectUri = bundle.get(Intent.EXTRA_STREAM);
            if (objectUri != null) {
               uri = (Uri) objectUri;
            }
         }
      }
      if (uri != null) {
         handleUri(uri);
      }

      // if token is not expired, then skip the login module
      if (!UserUtility.getInstance(getApplicationContext()).isTokenExpired()) {
         skipLogin();
      } else {
         // temporarily prune complete work on every login to ensure our unique work is rescheduled
         WorkManager.getInstance(getApplicationContext()).pruneWork();
         application.stopLocationService();
      }

      // no title bar
      setContentView(R.layout.activity_login);
      hideKeyboardOnClick(findViewById(R.id.login));

      TextView appName = findViewById(R.id.mage);
      appName.setTypeface(Typeface.createFromAsset(getAssets(), "fonts/GondolaMage-Regular.otf"));

      ((TextView) findViewById(R.id.login_version)).setText("App Version: " + preferences.getString(getString(R.string.buildVersionKey), "NA"));

      mServerURL = findViewById(R.id.server_url);

      String serverUrl = preferences.getString(getString(R.string.serverURLKey), getString(R.string.serverURLDefaultValue));
      if (StringUtils.isEmpty(serverUrl)) {
         changeServerURL();
         return;
      }

      findViewById(R.id.server_url).setOnClickListener(new View.OnClickListener() {
         @Override
         public void onClick(View v) {
            changeServerURL();
         }
      });

      mServerURL.setText(serverUrl);

      // Setup login based on last api pull
      configureLogin();

      viewModel = new ViewModelProvider(this).get(LoginViewModel.class);
      viewModel.getApiStatus().observe(this, valid -> observeApi());

      viewModel.getAuthenticationState().observe(this, this::observeAuthenticationState);
      viewModel.getAuthenticationStatus().observe(this, this::observeAuthentication);
      viewModel.getAuthorizationStatus().observe(this, this::observeAuthorization);

      viewModel.api(serverUrl);
   }

   @Override
   public void onBackPressed() {
      if (mContinueSession) {
         // In this case the activity stack was preserved. Don't allow the user to go back to an activity without logging in.
         // Since this is the application entry point, assume back means go home.
         Intent intent = new Intent(Intent.ACTION_MAIN);
         intent.addCategory(Intent.CATEGORY_HOME);
         intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
         startActivity(intent);
         return;
      }

      super.onBackPressed();
   }


   private void observeApi() {
      configureLogin();
   }

   private void observeAuthenticationState(LoginViewModel.AuthenticationState state) {
      if (state == LoginViewModel.AuthenticationState.LOADING) {
         findViewById(R.id.login_status).setVisibility(View.VISIBLE);
         findViewById(R.id.login_form).setVisibility(View.GONE);
      } else if (state == LoginViewModel.AuthenticationState.ERROR) {
         findViewById(R.id.login_status).setVisibility(View.GONE);
         findViewById(R.id.login_form).setVisibility(View.VISIBLE);
      }
   }

   private void observeAuthentication(Authentication authentication) {
      if (authentication == null) return;

      AuthenticationStatus status = authentication.getStatus();

      if (status.getStatus() == AuthenticationStatus.Status.SUCCESSFUL_AUTHENTICATION) {
         viewModel.authorize(authentication.getStrategy(), status.getToken(), status);
      } else if (status.getStatus() == AuthenticationStatus.Status.DISCONNECTED_AUTHENTICATION) {
         new AlertDialog.Builder(this)
                 .setTitle("Disconnected Login")
                 .setMessage("You are logging into MAGE in disconnected mode.  You must re-establish a connection in order to push and pull information to and from your server.")
                 .setPositiveButton(android.R.string.ok, (dialog, which) -> {
                    loginComplete(false);
                    dialog.dismiss();
                 })
                 .setCancelable(false)
                 .show();
      } else if (status.getStatus() == AuthenticationStatus.Status.ACCOUNT_CREATED) {
         String message = status.getMessage();
         if (message == null) {
            message = "Please contact a MAGE administrator to activate your account.";
         }

         new AlertDialog.Builder(this)
                 .setTitle("Account Created")
                 .setMessage(message)
                 .setPositiveButton(android.R.string.ok, null)
                 .show();
      } else {
         new AlertDialog.Builder(this)
                 .setTitle("Signin Failed")
                 .setMessage(status.getMessage())
                 .setPositiveButton(android.R.string.ok, null)
                 .show();
      }
   }

   private void observeAuthorization(LoginViewModel.Authorization authorization) {
      if (authorization == null) return;

      AuthorizationStatus.Status status = authorization.getStatus().getStatus();
      if (status == AuthorizationStatus.Status.SUCCESSFUL_AUTHORIZATION) {
         loginComplete(authorization.getUserChanged());
      } else if (status == AuthorizationStatus.Status.FAILED_AUTHORIZATION) {
         new AlertDialog.Builder(this)
                 .setTitle("Registration Sent")
                 .setMessage(R.string.device_registered_text)
                 .setPositiveButton(android.R.string.ok, null)
                 .show();
      } else if (status == AuthorizationStatus.Status.INVALID_SERVER) {
         new AlertDialog.Builder(this)
                 .setTitle("Application Compatibility Error")
                 .setMessage("This app is not compatible with this server. Please update your context or talk to your MAGE administrator.")
                 .setPositiveButton(android.R.string.ok, null)
                 .show();
      } else {
         new AlertDialog.Builder(this)
                 .setTitle("Signin Failed")
                 .setMessage(authorization.getStatus().getMessage())
                 .setPositiveButton(android.R.string.ok, null)
                 .show();
      }
   }

   private void configureLogin() {
      FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();

      Map<String, JSONObject> strategies = new TreeMap<>();

      // TODO marshal authentication strategies to POJOs with Jackson
      JSONObject authenticationStrategies = PreferenceHelper.getInstance(getApplicationContext()).getAuthenticationStrategies();
      Iterator<String> iterator = authenticationStrategies.keys();
      while (iterator.hasNext()) {
         String strategyKey = iterator.next();
         try {
            JSONObject strategy = (JSONObject) authenticationStrategies.get(strategyKey);
            if ("local".equals(strategyKey)) {
               strategy.putOpt("type", strategyKey);
            }

            strategies.put(strategyKey, strategy);
         } catch (JSONException e) {
            e.printStackTrace();
         }
      }

      if (strategies.size() > 1 && strategies.containsKey("local")) {
         findViewById(R.id.or).setVisibility(View.VISIBLE);
      } else {
         findViewById(R.id.or).setVisibility(View.GONE);
      }

      findViewById(R.id.google_login_button).setVisibility(View.GONE);
      for (final Map.Entry<String, JSONObject> entry : strategies.entrySet()) {
         String authenticationName = entry.getKey();
         String authenticationType = entry.getValue().optString("type");

         if (getSupportFragmentManager().findFragmentByTag(authenticationName) != null) continue;

         if ("local".equals(authenticationName)) {
            Fragment loginFragment = MageLoginFragment.Companion.newInstance(entry.getKey(), entry.getValue());
            transaction.add(R.id.local_auth, loginFragment, authenticationName);
         } else if ("oauth".equals(authenticationType)) {
            Fragment loginFragment = IdpLoginFragment.Companion.newInstance(entry.getKey(), entry.getValue());
            transaction.add(R.id.third_party_auth, loginFragment, authenticationName);
         } else if ("saml".equals(authenticationType)) {
            Fragment loginFragment = IdpLoginFragment.Companion.newInstance(entry.getKey(), entry.getValue());
            transaction.add(R.id.third_party_auth, loginFragment, authenticationName);
         } else if ("ldap".equals(authenticationType)) {
            Fragment loginFragment = LdapLoginFragment.Companion.newInstance(entry.getKey(), entry.getValue());
            transaction.add(R.id.third_party_auth, loginFragment, authenticationName);
         }
      }

      // Remove authentication fragments that have been removed from server
      for (Fragment fragment: getSupportFragmentManager().getFragments()) {
         if (!strategies.keySet().contains(fragment.getTag())) {
            transaction.remove(fragment);
         }
      }

      transaction.commit();
   }

   /**
    * Hides keyboard when clicking elsewhere
    *
    * @param view
    */
   private void hideKeyboardOnClick(View view) {
      // Set up touch listener for non-text box views to hide keyboard.
      if (!(view instanceof EditText) && !(view instanceof Button)) {
         view.setOnTouchListener(new OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
               InputMethodManager inputMethodManager = (InputMethodManager) getSystemService(Activity.INPUT_METHOD_SERVICE);
               if (getCurrentFocus() != null) {
                  inputMethodManager.hideSoftInputFromWindow(getCurrentFocus().getWindowToken(), 0);
               }
               return false;
            }
         });
      }

      // If a layout container, iterate over children and seed recursion.
      if (view instanceof ViewGroup) {
         for (int i = 0; i < ((ViewGroup) view).getChildCount(); i++) {
            View innerView = ((ViewGroup) view).getChildAt(i);
            hideKeyboardOnClick(innerView);
         }
      }
   }

   public void changeServerURL() {
      Intent intent = new Intent(this, ServerUrlActivity.class);
      startActivity(intent);
      finish();
   }

   /**
    * Handle the Uri used to launch MAGE
    *
    * @param uri
    */
   private void handleUri(Uri uri) {

      // Attempt to get a local file path
      String openPath = MediaUtility.getPath(this, uri);

      // If not a local or temporary file path, copy the file to cache
      // Cannot pass this to another activity to handle as the URI might
      // become invalid between now and then.  Copy it now
      if (openPath == null || MediaUtility.isTemporaryPath(openPath)) {
         CacheUtils.copyToCache(this, uri, openPath);
      } else {
         // Else, store the path to pass to further intents
         mOpenFilePath = openPath;
      }
   }

   /**
    * Fired when user clicks signup
    */
   public void signup(View view) {
      Intent intent;
      if (Compatibility.Companion.isServerVersion5(getApplicationContext())) {
         intent = new Intent(getApplicationContext(), SignupActivity_server5.class);

      } else {
         intent = new Intent(getApplicationContext(), SignupActivity.class);
      }

      startActivity(intent);
      finish();
   }

   private void loginComplete(Boolean userChanged) {
      boolean preserveActivityStack = !userChanged && mContinueSession;
      startNextActivityAndFinish(preserveActivityStack);
   }

   public void startNextActivityAndFinish(boolean preserveActivityStack) {
      // Continue session if there are other activities on the stack
      if (preserveActivityStack && !isTaskRoot()) {
         // We are going to return user to the app where they last left off,
         // make sure to start up MAGE services
         application.onLogin();

         // TODO look at refreshing the event here...
      } else {
         SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
         boolean showDisclaimer = sharedPreferences.getBoolean(getString(R.string.serverDisclaimerShow), false);

         Intent intent = showDisclaimer ?
                 new Intent(getApplicationContext(), DisclaimerActivity.class) :
                 new Intent(getApplicationContext(), EventsActivity.class);

         // If launched with a local file path, save as an extra
         if (mOpenFilePath != null) {
            intent.putExtra(LandingActivity.EXTRA_OPEN_FILE_PATH, mOpenFilePath);
         }

         startActivity(intent);
      }

      finish();
   }

   public void skipLogin() {
      Intent intent;

      SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
      boolean disclaimerAccepted = sharedPreferences.getBoolean(getString(R.string.disclaimerAcceptedKey), false);
      if (disclaimerAccepted) {
         Event event = null;
         try {
            User user = UserHelper.getInstance(getApplicationContext()).readCurrentUser();
            event = user.getCurrentEvent();
         } catch (UserException e) {
            e.printStackTrace();
         }

         intent = event == null ?
                 new Intent(getApplicationContext(), EventsActivity.class) :
                 new Intent(getApplicationContext(), LandingActivity.class);
      } else {
         intent = new Intent(getApplicationContext(), DisclaimerActivity.class);
      }

      // If launched with a local file path, save as an extra
      if (mOpenFilePath != null) {
         intent.putExtra(LandingActivity.EXTRA_OPEN_FILE_PATH, mOpenFilePath);
      }

      startActivity(intent);
      finish();
   }

   @Override
   protected void onResume() {
      super.onResume();

      if (getIntent().getBooleanExtra("LOGOUT", false)) {
         application.onLogout(true, null);
      }
   }

   private void showSessionExpiredDialog() {
      AlertDialog dialog = new AlertDialog.Builder(this)
              .setTitle("Session Expired")
              .setCancelable(false)
              .setMessage("We apologize, but it looks like your MAGE session has expired.  Please login and we will take you back to what you were doing.")
              .setPositiveButton(android.R.string.ok, null).create();

      dialog.setCanceledOnTouchOutside(false);

      dialog.show();
   }
}

File: mage/src/main/java/mil/nga/giat/mage/login/LoginActivity.kt
package mil.nga.giat.mage.login

import android.app.Activity
import android.content.DialogInterface
import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import android.os.Bundle
import android.util.Log
import androidx.activity.compose.setContent
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.work.WorkManager
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import mil.nga.giat.mage.LandingActivity
import mil.nga.giat.mage.MageApplication
import mil.nga.giat.mage.R
import mil.nga.giat.mage.R.xml
import mil.nga.giat.mage.cache.CacheUtils
import mil.nga.giat.mage.compat.server5.login.SignupActivityServer5
import mil.nga.giat.mage.contact.ContactDialog
import mil.nga.giat.mage.data.datasource.user.UserLocalDataSource
import mil.nga.giat.mage.database.dao.MageSqliteOpenHelper
import mil.nga.giat.mage.database.model.event.Event
import mil.nga.giat.mage.di.TokenProvider
import mil.nga.giat.mage.disclaimer.DisclaimerActivity
import mil.nga.giat.mage.event.EventsActivity
import mil.nga.giat.mage.login.AuthenticationStatus.AccountCreated
import mil.nga.giat.mage.login.AuthenticationStatus.Offline
import mil.nga.giat.mage.login.AuthorizationStatus.FailAuthentication
import mil.nga.giat.mage.login.AuthorizationStatus.FailAuthorization
import mil.nga.giat.mage.login.AuthorizationStatus.FailInvalidServer
import mil.nga.giat.mage.login.LoginViewModel.Authentication
import mil.nga.giat.mage.login.LoginViewModel.Authorization
import mil.nga.giat.mage.login.idp.IdpLoginActivity
import mil.nga.giat.mage.map.cache.CacheProvider
import mil.nga.giat.mage.sdk.Compatibility.Companion.isServerVersion5
import mil.nga.giat.mage.sdk.preferences.PreferenceHelper
import mil.nga.giat.mage.sdk.utils.MediaUtility
import mil.nga.giat.mage.ui.login.LoginScreen
import mil.nga.giat.mage.utils.IntentConstants
import org.apache.commons.lang3.StringUtils
import javax.inject.Inject

@AndroidEntryPoint
class LoginActivity : AppCompatActivity() {
   @Inject lateinit var application: MageApplication
   @Inject lateinit var preferences: SharedPreferences
   @Inject lateinit var tokenProvider: TokenProvider
   @Inject lateinit var userLocalDataSource: UserLocalDataSource
   @Inject lateinit var cacheProvider: CacheProvider

   private lateinit var viewModel: LoginViewModel

   private var mOpenFilePath: String? = null
   private var mContinueSession = false

   private lateinit var idpLoginLauncher: ActivityResultLauncher<Intent>

   public override fun onCreate(savedInstanceState: Bundle?) {
      super.onCreate(savedInstanceState)

      val serverUrl = preferences.getString(getString(R.string.serverURLKey), getString(R.string.serverURLDefaultValue))!!

      //if serverUrl is not populated, then the user needs to specify one prior to login
      if (StringUtils.isEmpty(serverUrl)) {
         changeServerURL()
      } else {
         // if token is not expired, then skip the login module
         if (!tokenProvider.isExpired()) {
            skipLogin()
         } else {
            // temporarily prune complete work on every login to ensure our unique work is rescheduled
            WorkManager.getInstance(applicationContext).pruneWork()
            application.stopLocationService()

            val intent = intent
            mContinueSession = getIntent().getBooleanExtra(EXTRA_CONTINUE_SESSION, false)

            val continueSessionWhileUsing =
               getIntent().getBooleanExtra(EXTRA_CONTINUE_SESSION_WHILE_USING, false)

            intent.removeExtra(EXTRA_CONTINUE_SESSION_WHILE_USING)
            if (continueSessionWhileUsing && savedInstanceState == null) {
               showSessionExpiredDialog()
            }

            val version = "App Version: " + preferences.getString(getString(R.string.buildVersionKey), "NA")

            // IMPORTANT: load the configuration from preferences files and server
            val preferenceHelper = PreferenceHelper.getInstance(applicationContext)
            preferenceHelper.initialize(false, xml::class.java)

            // check if the database needs to be upgraded, and if so log them out
            if (MageSqliteOpenHelper.DATABASE_VERSION != preferences.getInt(
                  resources.getString(R.string.databaseVersionKey), 0)) {
               application.onLogout(true)
            } else if (intent.getBooleanExtra("LOGOUT", false)) {
               application.onLogout(true)
            }

            preferences.edit().putInt(getString(R.string.databaseVersionKey), MageSqliteOpenHelper.DATABASE_VERSION).apply()

            //check Google Play version - a minimum version is required to use the app
            checkGooglePlay()

            // Handle when MAGE was launched with a Uri (such as a local or remote cache file)
            var uri = intent.data
            if (uri == null) {
               val bundle = intent.extras
               if (bundle != null) {
                  val objectUri = bundle[Intent.EXTRA_STREAM]
                  if (objectUri != null) {
                     uri = objectUri as Uri?
                  }
               }
            }
            uri?.let { handleUri(it) }

            //launches web login page hosted outside the app for IDP auth (SAML, OPENIDCONNECT, OAUTH)
            val onIdpLoginClick = { authType: ServerAuthTypes ->
               val idpLoginIntent = IdpLoginActivity.intent(this, serverUrl, authType)
               idpLoginLauncher.launch(idpLoginIntent)
            }

            //callback handler from IdpLoginActivity
            idpLoginLauncher =
               registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
                  if (result.resultCode == Activity.RESULT_OK) {
                     val idpResponseToken = result.data?.getStringExtra(IdpLoginActivity.EXTRA_IDP_TOKEN)
                     val upperCaseStrategyName = result.data?.getStringExtra(IdpLoginActivity.EXTRA_IDP_STRATEGY)?.uppercase()

                     val authType = ServerAuthTypes.entries.find { it.name == upperCaseStrategyName }
                     if (authType != null) {
                        viewModel.authorize(authType, idpResponseToken ?: "")
                     } else {
                        Log.e(LOG_NAME, "IDP Login Failed or Cancelled")
                        processAuthenticationResult(Authentication(ServerAuthTypes.SAML, AuthenticationStatus.Failure(0, "Login failed or cancelled")))
                     }
                  } else {
                     Log.e(LOG_NAME, "IDP Login Failed or Cancelled")
                     processAuthenticationResult(Authentication(ServerAuthTypes.SAML, AuthenticationStatus.Failure(0, "Login failed or cancelled")))
                  }
               }

            viewModel = ViewModelProvider(this).get(LoginViewModel::class.java)

            //retrieve previously stored authStrategies JSON and launch call to refresh it
            viewModel.availableLoginTypesJSON.value = PreferenceHelper.getInstance(applicationContext).authenticationStrategies
            viewModel.checkApi(serverUrl)

            //server authentication for LOCAL and LDAP auth
            val authenticate = {
               authType: ServerAuthTypes, userId: String, pwd: String ->
                  viewModel.authenticate(authType, userId.lowercase(), pwd)
            }

            setContent {
               LoginScreen(
                  viewModel.availableLoginTypesJSON,
                  viewModel.showProgressSpinner,
                  onServerUrlClick = { changeServerURL(true) },
                  onLoginClick = authenticate,
                  onIdpLoginClick = onIdpLoginClick,
                  onSignUpClick = { signup() },
                  serverUrl = serverUrl,
                  version = version
               )
            }

            //collector for authentication and authorization events
            collectAuthEvents()
         }
      }
   }

   //collect auth events fired from LoginViewModel
   private fun collectAuthEvents() {
      lifecycleScope.launch {
         //repeatOnLifecycle ensures collection stops when view is paused/stopped, and restarts when resumed
         repeatOnLifecycle(Lifecycle.State.STARTED) {

            //monitor auth state to control display of progress spinner
            launch {
               viewModel.authenticationProcessState.collectLatest { state ->
                  updateProgressSpinner(state)
               }
            }

            //process the authentication result event
            launch {
               viewModel.authenticationResultEvents.collectLatest { authentication ->
                  processAuthenticationResult(authentication)
               }
            }

            //process the authorization result event
            launch {
               viewModel.authorizationResultEvents.collectLatest { authorization ->
                  processAuthorizationResult(authorization)
               }
            }

            //process the "api" result event to update available login options based on server config
            launch {
               viewModel.apiStatusSuccessEvent.collectLatest {
                  updateLoginOptions()
               }
            }
         }
      }
   }

   //check google play services version
   private fun checkGooglePlay() {
      val isGooglePlayServicesAvailable =
         GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(applicationContext)
      if (isGooglePlayServicesAvailable != ConnectionResult.SUCCESS) {
         if (GoogleApiAvailability.getInstance()
               .isUserResolvableError(isGooglePlayServicesAvailable)
         ) {
            val dialog = GoogleApiAvailability.getInstance()
               .getErrorDialog(this, 1, isGooglePlayServicesAvailable)
            dialog?.setOnCancelListener { dialog1: DialogInterface ->
               dialog1.dismiss()
               finish()
            }
            dialog?.show()
         } else {
            AlertDialog.Builder(this).setTitle("Google Play Services")
               .setMessage("Google Play Services is not installed, or needs to be updated.  Please update Google Play Services before continuing.")
               .setPositiveButton(
                  android.R.string.ok
               ) { dialog, _ ->
                  dialog.dismiss()
                  finish()
               }.show()
         }
      }
   }

   override fun onBackPressed() {
      if (mContinueSession) {
         // In this case the activity stack was preserved. Don't allow the user to go back to an activity without logging in.
         // Since this is the application entry point, assume back means go home.
         val intent = Intent(Intent.ACTION_MAIN)
         intent.addCategory(Intent.CATEGORY_HOME)
         intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
         startActivity(intent)
         return
      }
      super.onBackPressed()
   }

   private fun updateProgressSpinner(state: AuthenticationState) {
      if (state === AuthenticationState.LOADING) {
         viewModel.showProgressSpinner.value = true
      } else {
         viewModel.showProgressSpinner.value = false
      }
   }

   private fun processAuthenticationResult(authentication: Authentication?) {
      if (authentication == null) return
      when (val status = authentication.status){
         is AuthenticationStatus.Success -> {
            val token = status.token
            viewModel.authorize(authentication.authType, token)
         }
         is AccountCreated -> {
            val message = status.message
            val dialog = ContactDialog(this, (preferences), "Account Created", message)
            dialog.setAuthenticationStrategy(authentication.authType.name)
            dialog.show(null)
         }
         is Offline -> {
            val message = status.message
            val dialog = ContactDialog(this, (preferences), "Sign in Failed", message)
            dialog.setAuthenticationStrategy(authentication.authType.name)
            dialog.show { workOffline: Boolean ->
               if (workOffline) {
                  loginComplete(false)
               }
               viewModel.completeOffline(workOffline)
            }
         }
         is AuthenticationStatus.Failure -> {
            val message = status.message
            val dialog = ContactDialog(this, (preferences), "Sign in Failed", message)
            dialog.setAuthenticationStrategy(authentication.authType.name)
            dialog.show(null)
         }
      }
   }

   private fun processAuthorizationResult(authorization: Authorization?) {
      if (authorization == null) return
      when (val status = authorization.status) {
         is AuthorizationStatus.Success -> {
            loginComplete(status.sessionChanged)
         }
         is FailAuthorization -> {
            val dialog = ContactDialog(
               this,
               preferences,
               "Registration Sent",
               getString(R.string.device_registered_text)
            )
            val user = status.user
            if (user != null) {
               dialog.username = user.username
            }
            dialog.show(null)
         }
         is FailInvalidServer -> {
            val dialog = ContactDialog(
               this,
               preferences,
               "Application Compatibility Error",
               "MAGE is not compatible with this server, please ensure your application is up to date or contact your MAGE administrator."
            )
            dialog.show(null)
         }
         is FailAuthentication -> {
            val message = status.message
            val dialog = ContactDialog(this, preferences, "Sign-in Failed", message)
            val user = status.user
            if (user != null) {
               dialog.username = user.username
            }
            dialog.show(null)
         }
      }
   }

   private fun updateLoginOptions() {
      viewModel.availableLoginTypesJSON.value = PreferenceHelper.getInstance(applicationContext).authenticationStrategies
   }


   private fun changeServerURL(launchedFromButtonClick: Boolean = false) {
      val intent = Intent(this, ServerUrlActivity::class.java)
      intent.putExtra(IntentConstants.LAUNCHED_FROM_BUTTON_CLICK, launchedFromButtonClick)
      startActivity(intent)
      finish()
   }

   /**
    * Handle the Uri used to launch MAGE
    *
    * @param uri
    */
   private fun handleUri(uri: Uri) {
      // Attempt to get a local file path
      val openPath = MediaUtility.getPath(this, uri)

      // If not a local or temporary file path, copy the file to cache
      // Cannot pass this to another activity to handle as the URI might
      // become invalid between now and then.  Copy it now
      if (openPath == null || MediaUtility.isTemporaryPath(openPath)) {
         CoroutineScope(Dispatchers.IO).launch {
            CacheUtils(applicationContext, cacheProvider).copyToCache(uri, openPath)
         }

      } else {
         // Else, store the path to pass to further intents
         mOpenFilePath = openPath
      }
   }

   /**
    * Fired when user clicks signup
    */
   fun signup() {
      val intent = if (isServerVersion5(applicationContext)) {
         Intent(applicationContext, SignupActivityServer5::class.java)
      } else {
         Intent(applicationContext, SignupActivity::class.java)
      }
      startActivity(intent)
      finish()
   }

   private fun loginComplete(userChanged: Boolean) {
      val preserveActivityStack = !userChanged && mContinueSession
      startNextActivityAndFinish(preserveActivityStack)
   }

   private fun startNextActivityAndFinish(preserveActivityStack: Boolean) {
      // Continue session if there are other activities on the stack
      if (preserveActivityStack && !isTaskRoot) {
         // We are going to return user to the app where they last left off,
         // make sure to start up MAGE services
         application.onLogin()

         // TODO look at refreshing the event here...
      } else {
         val showDisclaimer = preferences.getBoolean(getString(R.string.serverDisclaimerShow), false)
         val intent = if (showDisclaimer) Intent(
            applicationContext,
            DisclaimerActivity::class.java
         ) else Intent(
            applicationContext, EventsActivity::class.java
         )

         // If launched with a local file path, save as an extra
         if (mOpenFilePath != null) {
            intent.putExtra(LandingActivity.EXTRA_OPEN_FILE_PATH, mOpenFilePath)
         }
         startActivity(intent)
      }
      finish()
   }

   private fun skipLogin() {
      val intent: Intent
      val disclaimerAccepted = preferences.getBoolean(getString(R.string.disclaimerAcceptedKey), false)
      if (disclaimerAccepted) {
         var event: Event? = null
         val user = userLocalDataSource.readCurrentUser()
         if (user != null) {
            event = user.currentEvent
         }
         intent =
            if (event == null) Intent(applicationContext, EventsActivity::class.java) else Intent(
               applicationContext, LandingActivity::class.java
            )
      } else {
         intent = Intent(applicationContext, DisclaimerActivity::class.java)
      }

      // If launched with a local file path, save as an extra
      if (mOpenFilePath != null) {
         intent.putExtra(LandingActivity.EXTRA_OPEN_FILE_PATH, mOpenFilePath)
      }
      startActivity(intent)
      finish()
   }

   override fun onResume() {
      super.onResume()
      if (intent.getBooleanExtra("LOGOUT", false)) {
         application.onLogout(true)
      }
   }

   private fun showSessionExpiredDialog() {
      val dialog = AlertDialog.Builder(this)
         .setTitle("Session Expired")
         .setCancelable(false)
         .setMessage("We apologize, but it looks like your MAGE session has expired.  Please login and we will take you back to what you were doing.")
         .setPositiveButton(android.R.string.ok, null).create()
      dialog.setCanceledOnTouchOutside(false)
      dialog.show()
   }

   companion object {
      private val LOG_NAME = LoginActivity::class.java.name
      const val EXTRA_CONTINUE_SESSION = "CONTINUE_SESSION"
      const val EXTRA_CONTINUE_SESSION_WHILE_USING = "CONTINUE_SESSION_WHILE_USING"
   }
}